2023-11-16-Rust-1.74.0.md (#130)
rust-lang-ru/translations: ru/announcements/2023-11-16-Rust-1.74.0.md

---
layout: post
title: 'Rust 1.74.0: конфигурация проверки в Cargo, авторизация для приватных реестров, проекции возвращаемых типов'
author: The Rust Release Team
release: 'true'
---

Команда Rust рада сообщить о новой версии языка — 1.74.0. Rust — это язык программирования, позволяющий каждому создавать надёжное и эффективное программное обеспечение.

Если у вас есть предыдущая версия Rust, установленная через `rustup` то для обновления до версии 1.74.0 вам достаточно выполнить команду:

```console
rustup update stable
```

Если у вас ещё не установлен <code>rustup</code>, вы можете установить его с <a>соответствующей страницы</a> нашего веб-сайта, а также посмотреть [подробные примечания к выпуску](https://github.com/rust-lang/rust/releases/tag/1.74.0) на GitHub.

Если вы хотите помочь нам протестировать будущие выпуски, вы можете использовать beta (`rustup default beta`) или nightly (`rustup default nightly`) канал. Пожалуйста, [сообщайте](https://github.com/rust-lang/rust/issues/new/choose) обо всех встреченных вами ошибках.

## Что стабилизировано в 1.74.0

> ### Проверка конфигурации при помощи Cargo

Reviewer: Не "Проверка конфигурации", а, наоборот, "Настройка проверок" (или не "проверок", а как там lints принято переводить).

Reviewer: @funkill Слово "конфигурации" тоже надо убрать. Линты проверяют код, а не конфигурацию чего бы то ни было.

> Как предложено в [RFC 3389](https://rust-lang.github.io/rfcs/3389-manifest-lint.html), файл `Cargo.toml` теперь поддерживает таблицу `[lints]` для настройки уровня логгирования ошибок (forbid, deny, warn, allow) для проверок от компилятора и других инструментов. Таким образом, вместо установки `RUSTFLAGS` в `-F`/`-D`/`-W`/`-A`, что повлияло бы на всю сборку, или используя атрибуты уровня крейта, такие как:

Reviewer 2: ```suggestion
Как предложено в [RFC 3389](https://rust-lang.github.io/rfcs/3389-manifest-lint.html), файл `Cargo.toml` теперь поддерживает таблицу `[lints]` для настройки уровня сообщений (forbid, deny, warn, allow) для проверок от компилятора и других инструментов. Таким образом, вместо установки `RUSTFLAGS` в `-F`/`-D`/`-W`/`-A`, что повлияло бы на всю сборку, или используя атрибуты уровня крейта, такие как:
```

```rust
#![forbid(unsafe_code)]
#![deny(clippy::enum_glob_use)]
```

> Теперь вы можете указать их в манифесте пакета чтобы их обработал Cargo:

Reviewer 2: ```suggestion
Теперь вы можете указать их в манифесте пакета, чтобы их обработал Cargo:
```

```toml
[lints.rust]
unsafe_code = "forbid"

[lints.clippy]
enum_glob_use = "deny"
```

Также они могут быть настроены в таблице `[workspace.lints]`, затем унаследованы `[lints] workspace = true`, как и многие другие параметры рабочей области. Cargo также будет отслеживать изменения в этих настройках при принятии решения о том, какие крейты необходимо пересобрать.

Для получения дополнительной информации смотрите разделы [`lints`] и [`workspace.lints`] справочного руководства по Cargo.

### Авторизация в Cargo Registry

В этот выпуск включены ещё две связанные с Cargo функции: провайдеры учётных данных и аутентификация для приватных реестров.

> Провайдеры авторизации позволяют настраивать способ получения Cargo учётных данных для реестра. Встроенные провайдеры включены для безопасных секретных хранилищей, специфичных для конкретной операционной системы, в Linux, macOS и Windows. Кроме того, для поддержки произвольных методов хранения или генерации токенов могут быть созданы сторонние провайдеры. Использование защищённых провайдеров учётных данных снижает риск утечки токенов реестра.

Reviewer 2: ```suggestion
Провайдеры авторизации позволяют настраивать способ получения Cargo учётных данных для реестра. Встроенные провайдеры включены для безопасных секретных хранилищ, специфичных для конкретной операционной системы, в Linux, macOS и Windows. Кроме того, для поддержки произвольных методов хранения или генерации токенов могут быть созданы сторонние провайдеры. Использование защищённых провайдеров учётных данных снижает риск утечки токенов реестра.
```

Дополнительно реестры теперь могут требовать аутентификацию для всех операций, а не только для публикации. Это позволяет приватным реестрам предлагать более безопасное размещение крейтов. Использование приватных реестров требует настройки провайдера авторизации.

> Для большей информации смотрите [документацию Cargo](https://doc.rust-lang.org/beta/cargo/reference/registry-authentication.html).

Reviewer 2: ```suggestion
Более подробную информацию смотрите в [документации Cargo](https://doc.rust-lang.org/beta/cargo/reference/registry-authentication.html).
```

### Проекции в непрозрачных возвращаемых типах

> Если вы когда-либо получали сообщение об ошибке вроде "return type cannot contain a projection or `Self` that references lifetimes from a parent scope,", то теперь вы можете быть спокойны! Компилятор позволяет упоминать `Self` и связанные с ним типы в непрозрачных возвращаемых типах, таких как `async fn` и `-> impl Trait`. Это тип особенностей, которые приближают нас к *ожиданиям* того, как оно должно работать, даже если вы понятия не имеете о таком жаргонизме, как "проекция".

Reviewer 2: ```suggestion
Если вы когда-либо получали сообщение об ошибке вроде "return type cannot contain a projection or `Self` that references lifetimes from a parent scope,", то теперь вы можете быть спокойны! Компилятор позволяет упоминать `Self` и связанные с ним типы в непрозрачных возвращаемых типах, таких как `async fn` и `-> impl Trait`. Благодаря улучшениям вроде этого Rust лучше соответствует вашим *интуитивным ожиданиям* от него, даже если вы понятия не имеете о таких элементах технического жаргона, как "проекция".
```

> Эта функциональность была не стабильной потому что изначально ее реализация неправильно обрабатывала захваченные времена жизни, и как только это было исправлено, было взято время убедиться, что она исправна. Для получения более подробной технической информации смотрите [запрос на стабилизацию], в котором описаны все примеры, которые теперь разрешены:

Reviewer: Вместо "была не стабильной" лучше "имела статус нестабильной" или как-то так, чтобы было ясно, что это слово с конкретным техническим значением.
Нужна запятая перед "потому что".
Лучше не "исправна", а "не содержит ошибок, ведущих к неопределённом поведению" (т. к. sound тоже термин именно с таким смыслом).

Reviewer 2: ```suggestion
Эта функциональность имела статус не стабильной, потому что изначально ее реализация неправильно обрабатывала захваченные времена жизни, и как только это было исправлено, было взято время убедиться, что она не содержит ошибок, ведущих к неопределённом поведению. Для получения более подробной технической информации смотрите [запрос на стабилизацию], в котором описаны все примеры, которые теперь разрешены:
```

- [`impl<T: Clone, const N: usize> From<&mut [T; N]> for Vec<T>`](https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#impl-From%3C%26mut+%5BT;+N%5D%3E-for-Vec%3CT,+Global%3E)
- [`impl<T, const N: usize> From<[T; N]> for Arc<[T]>`](https://doc.rust-lang.org/stable/std/sync/struct.Arc.html#impl-From%3C%5BT;+N%5D%3E-for-Arc%3C%5BT%5D,+Global%3E)
- [`impl<T, const N: usize> From<[T; N]> for Rc<[T]>`](https://doc.rust-lang.org/stable/std/rc/struct.Rc.html#impl-From%3C%5BT;+N%5D%3E-for-Rc%3C%5BT%5D,+Global%3E)

Следующие API теперь можно использовать в контексте `const`:

- [`core::mem::transmute_copy`](https://doc.rust-lang.org/stable/std/mem/fn.transmute_copy.html)
- [`str::is_ascii`](https://doc.rust-lang.org/stable/std/primitive.str.html#method.is_ascii)
- [`[u8]::is_ascii`](https://doc.rust-lang.org/stable/std/primitive.slice.html#method.is_ascii)

### Замечания о совместимости

- Как было [сообщено ранее], Rust 1.74 повышает требования к платформам Apple. Теперь минимальные версии такие:
    - macOS: 10.12 Sierra (первый выпуск 2016)
    - iOS: 10 (первый выпуск 2016)
    - tvOS: 10 (первый выпуск 2016)

### Прочие изменения

> Проверьте всё, что изменилось в [Rust](https://github.com/rust-lang/rust/releases/tag/1.74.0), [Cargo](https://github.com/rust-lang/cargo/blob/master/CHANGELOG.md#cargo-174-2023-11-16) и [Clippy](https://github.com/rust-lang/rust-clippy/blob/master/CHANGELOG.md#rust-174).

Reviewer: Скорее "Взгляните на все изменения в". Но это мелочь.
